src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -61,55 +62,43 @@ impl ScalarFunc {
         Ok(Some(input.len() as i64))
     }
 
+    #[inline]
+    fn find_str(text: &str, pattern: &str) -> Option<i64> {
+        twoway::find_str(text, pattern).map(|i| text[..i].chars().count() as i64)
+    }
+
     #[inline]
     pub fn locate_2_args(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         let substr = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
         let s = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
-        if substr.is_empty() {
-            return Ok(Some(1));
-        }
-        if s.is_empty() {
-            return Ok(Some(0));
-        }
-        let substr = substr.to_lowercase();
-        let s = s.to_lowercase();
-        let result = match s.find(&substr) {
-            None => 0,
-            Some(i) => 1 + s[..i].chars().count() as i64,
-        };
-        Ok(Some(result))
+        Ok(Self::find_str(&s.to_lowercase(), &substr.to_lowercase())
+            .map(|i| 1 + i)
+            .or(Some(0)))
     }
 
     #[inline]
     pub fn locate_3_args(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         let substr = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
         let s = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
-        let mut pos = try_opt!(self.children[2].eval_int(ctx, row));
-
-        pos -= 1; // transfer 1-based argument to 0-based real index
-        if pos < 0 {
-            return Ok(Some(0));
-        }
-        let substr = substr.to_lowercase();
-        let s = s.to_lowercase();
-        if pos > (s.chars().count() as i64 - substr.chars().count() as i64) {
+        let pos = try_opt!(self.children[2].eval_int(ctx, row));
+        if pos < 1 {
             return Ok(Some(0));
         }
-        if substr.is_empty() {
-            return Ok(Some(pos + 1));
-        }
-
-        let slice = s
-            .char_indices()
-            .nth(pos as usize)
-            .map(|(offset, _)| &s[offset..]);
+        Ok(s.char_indices()
+            .map(|(i, _)| i)
+            .chain(iter::once(s.len()))
+            .nth(pos as usize - 1)
+            .map(|offset| {
+                Self::find_str(&s[offset..].to_lowercase(), &substr.to_lowercase())
+                    .map(|i| i + pos)
+                    .unwrap_or(0)
+            })
+            .or(Some(0)))
+    }
 
-        if let Some(slice) = slice {
-            if let Some(idx) = slice.find(&substr) {
-                return Ok(Some(1 + pos + slice[..idx].chars().count() as i64));
-            }
-        }
-        Ok(Some(0))
+    #[inline]
+    fn find_bytes(text: &[u8], pattern: &[u8]) -> Option<i64> {
+        twoway::find_bytes(text, pattern).map(|i| i as i64)
```

Comment:
Maybe this function can be just removed at all? It nearly encapsulates nothing.

Reply:
NP! Thx!